unsafe/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -173,18 +173,35 @@ public BytesToBytesMap(
   public Iterator<Location> iterator() {
     return new Iterator<Location>() {
 
-      private int nextPos = bitset.nextSetBit(0);
+      private int cur = 0;
+
+      private int pageCur = 0;
+
+      private MemoryBlock currentPage;
+
+      private long addr;
 
       @Override
       public boolean hasNext() {
-        return nextPos != -1;
+        return cur != size;
       }
 
       @Override
       public Location next() {
-        final int pos = nextPos;
-        nextPos = bitset.nextSetBit(nextPos + 1);
-        return loc.with(pos, 0, true);
+        if (currentPage == null) {
+          currentPage = dataPages.get(pageCur++);
+          addr = currentPage.getBaseOffset();
+        }
+        long keySize = PlatformDependent.UNSAFE.getLong(memoryManager.getPage(addr), addr);
+        if (keySize == -1L) {
```

Comment:
This branch was never taken in the unit tests but was exercised by my stress tester, which explains why I was able to find a bug even though the tests passed.  This is why I'm loving IntelliJ's code coverage metrics for this Tungsten stuff: insisting on 100% line coverage really helps to prevent bugs like this.